File: internal/audio/synthesizer.go
package audio

import (
	"log"
	"math"
	"sync"

	"github.com/robertpelloni/bqt/internal/ui"
)

// Synthesizer is the Go-native implementation of OmniSynthesizer.
// It conforms to the AudioNode interface and manages polyphony and basic synthesis.
type Voice struct {
	velocity float32
	phase    float64
}

type Synthesizer struct {
	mu          sync.Mutex
	polyphony   int
	activeNotes map[int]*Voice
	sampleRate  float64
}

// NewSynthesizer creates a new native Go synthesizer.
func NewSynthesizer() *Synthesizer {
	return &Synthesizer{
		polyphony:   8,
		activeNotes: make(map[int]*Voice),
		sampleRate:  44100.0,
	}
}

// Name returns the node name for the AudioGraph.
func (s *Synthesizer) Name() string {
	return "OmniSynthesizer"
}

// Polyphony returns the maximum number of concurrent voices.
func (s *Synthesizer) Polyphony() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.polyphony
}

// SetPolyphony sets the maximum number of concurrent voices.
func (s *Synthesizer) SetPolyphony(voices int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.polyphony != voices {
		s.polyphony = voices
		log.Printf("OmniSynthesizer Go: Polyphony set to %d", voices)
	}
}

// NoteOn registers a new active note.
func (s *Synthesizer) NoteOn(channel, note, velocity int) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.activeNotes) >= s.polyphony {
		// Ignore if max polyphony reached (no voice stealing implemented yet)
		log.Printf("OmniSynthesizer Go: Voice limit reached, ignoring NoteON")
		return
	}

	s.activeNotes[note] = &Voice{velocity: float32(velocity) / 127.0, phase: 0.0}

	// Post event to the unified BQt Event Loop for UI thread synchronization
	ui.GetEventLoop().Post(func() {
		log.Printf("OmniSynthesizer Go: Native Synthesis NoteON -> CH: %d NOTE: %d", channel, note)
	})
}

// NoteOff removes an active note.
func (s *Synthesizer) NoteOff(channel, note int) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.activeNotes, note)

	// Post event to the unified BQt Event Loop for UI thread synchronization
	ui.GetEventLoop().Post(func() {
		log.Printf("OmniSynthesizer Go: Native Synthesis NoteOFF -> CH: %d NOTE: %d", channel, note)
	})
}

// Process generates the audio buffer for all active voices and adds it to the output.
func (s *Synthesizer) Process(buffer []float32) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.activeNotes) == 0 {
		return // Nothing to synthesize
	}

	for i := 0; i < len(buffer); i++ {
		var sample float32 = 0.0

		// Render active notes as sine waves
		for note, voice := range s.activeNotes {
			// Midi note to frequency
			freq := 440.0 * math.Pow(2.0, float64(note-69)/12.0)
			sample += voice.velocity * float32(math.Sin(2.0*math.Pi*voice.phase))
			voice.phase += freq / s.sampleRate
			if voice.phase > 1.0 {
				voice.phase -= 1.0
			}
		}

		// Add to buffer (mix with existing audio)
		buffer[i] += sample

	}
}

File: rust/src/audio/synthesizer.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use super::audio_graph::AudioNode;

pub struct Voice {
    pub velocity: f32,
    pub phase: f64,
}

pub struct Synthesizer {
    polyphony: usize,
    active_notes: HashMap<usize, Voice>,
    sample_rate: f64,
}

pub struct OmniSynthesizer {
    inner: Arc<Mutex<Synthesizer>>,
}

impl OmniSynthesizer {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(Synthesizer {
                polyphony: 8,
                active_notes: HashMap::new(),
                sample_rate: 44100.0,
            })),
        }
    }

    pub fn polyphony(&self) -> usize {
        self.inner.lock().unwrap().polyphony
    }

    pub fn set_polyphony(&self, voices: usize) {
        let mut inner = self.inner.lock().unwrap();
        if inner.polyphony != voices {
            inner.polyphony = voices;
            println!("OmniSynthesizer Rust: Polyphony set to {}", voices);
        }
    }

    pub fn note_on(&self, channel: usize, note: usize, velocity: usize) {
        let mut inner = self.inner.lock().unwrap();
        if inner.active_notes.len() >= inner.polyphony {
            println!("OmniSynthesizer Rust: Voice limit reached, ignoring NoteON");
            return;
        }

        inner.active_notes.insert(note, Voice {
            velocity: (velocity as f32) / 127.0,
            phase: 0.0,
        });
        println!("OmniSynthesizer Rust: Native Synthesis NoteON -> CH: {} NOTE: {}", channel, note);
    }

    pub fn note_off(&self, channel: usize, note: usize) {
        let mut inner = self.inner.lock().unwrap();
        inner.active_notes.remove(&note);
        println!("OmniSynthesizer Rust: Native Synthesis NoteOFF -> CH: {} NOTE: {}", channel, note);
    }

    // Test helper
    pub fn active_notes_count(&self) -> usize {
        self.inner.lock().unwrap().active_notes.len()
    }
}

impl AudioNode for OmniSynthesizer {
    fn name(&self) -> &str {
        "OmniSynthesizer"
    }

    fn process(&mut self, buffer: &mut [f32]) {
        let mut inner = self.inner.lock().unwrap();

        if inner.active_notes.is_empty() {
            return;
        }

        let sample_rate = inner.sample_rate;

        for sample in buffer.iter_mut() {
            let mut mixed_sample = 0.0_f32;

            for (note, voice) in inner.active_notes.iter_mut() {
                let freq = 440.0 * 2.0_f64.powf((*note as f64 - 69.0) / 12.0);
                mixed_sample += voice.velocity * (2.0 * std::f64::consts::PI * voice.phase).sin() as f32;

                voice.phase += freq / sample_rate;
                if voice.phase > 1.0 {
                    voice.phase -= 1.0;
                }
            }

            *sample += mixed_sample;
        }
    }
}
